### PR TITLE
Add filecache support for ORC [databricks]

### LINKED ISSUE #8231: [FEA] Add filecache support to ORC scans
The filecache support added in #8212 currently only supports Parquet scans.  It would be nice if this also worked for ORC scans.

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #8231.  This adds filecache support to ORC scans.  In order to support file caching within ORC, we cannot leverage the ORC DataReader as we did before. The DataReader is responsible for loading the data, so therefore we need a custom version of that class that has the filecaching logic added to it.  This necessitated copying some code from ORC, especially for the Cloudera shims since they are on a much older version of ORC where the DataReader performs more responsibility than just loading data ranges.

This also eliminates some duplication between the 321cdh and 330cdh shims, as they're based on the same old ORC version, and thus large portions of the ORC shims apply to both.

Ran performance benchmarks before/after with filecache disabled, since we're changing the code paths for ORC reads even when filecache is disabled.  There was no measurable difference on NDS benchmark at SF=3k.

### CODE REVIEW

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Review by tgravescs — COMMENTED** [review]
one other comment otherwise looks good

File: sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/shims/GpuOrcDataReaderBase.scala [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    missMetric += 1
+    missSizeMetric += last.getLength
+    var cachedChannel: Option[SeekableByteChannel] = None
+    try {
+      while (cachedChannel.isEmpty &&
+          last.next != null &&
+          last.next.getOffset == currentEnd &&
+          last.next.getEnd - first.getOffset <= Int.MaxValue) {
+        val offset = baseOffset + last.next.getOffset
+        cachedChannel = FileCache.get.getDataRangeChannel(filePathString, offset,
```
[the remark this answers is not in the document]
Comment (jlowe) [reply sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/shims/GpuOrcDataReaderBase.scala:215]: This is not a double-lookup in practice.  The caller has alreadu done a cache lookup  on `last` for the first iteration of this loop, but this code does a cache lookup on `last.next`, i.e.: the next block, something we haven't tried yet.

**Comment by jlowe** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
Need to update premerge build to initialize and update submodules.

**Review by tgravescs — APPROVED** [review]

**Pull request merged**